File: frontend/src/types/customer.ts
/**
 * Customer Profile and Measurements - TypeScript types and Zod schemas
 * Story 1.3: Quản lý Hồ sơ & Số đo
 */

import { z } from "zod";

// ===== Validation Schemas =====

/**
 * Vietnamese phone number validation
 * Format: 0XXXXXXXXX (10-11 digits)
 */
const phoneSchema = z
  .string()
  .min(10, "Số điện thoại phải có ít nhất 10 số")
  .max(11, "Số điện thoại không được quá 11 số")
  .regex(/^0[0-9]{9,10}$/, "Số điện thoại không đúng định dạng (VD: 0901234567)");

/**
 * Measurement field validation with range
 */
const measurementFieldSchema = (min: number, max: number, fieldName: string) =>
  z
    .number()
    .min(min, `${fieldName} phải lớn hơn ${min}cm`)
    .max(max, `${fieldName} phải nhỏ hơn ${max}cm`)
    .positive(`${fieldName} phải là số dương`)
    .optional();

// ===== Measurement Schemas =====

/**
 * Schema for creating a new measurement set
 */
export const measurementCreateSchema = z.object({
  neck: measurementFieldSchema(20, 60, "Vòng cổ"),
  shoulder_width: measurementFieldSchema(30, 60, "Rộng vai"),
  bust: measurementFieldSchema(60, 180, "Vòng ngực"),
  waist: measurementFieldSchema(40, 150, "Vòng eo"),
  hip: measurementFieldSchema(60, 180, "Vòng mông"),
  top_length: measurementFieldSchema(40, 120, "Dài áo"),
  sleeve_length: measurementFieldSchema(30, 90, "Dài tay"),
  wrist: measurementFieldSchema(10, 30, "Vòng cổ tay"),
  height: measurementFieldSchema(100, 250, "Chiều cao"),
  weight: z
    .number()
    .min(30, "Cân nặng phải lớn hơn 30kg")
    .max(200, "Cân nặng phải nhỏ hơn 200kg")
    .positive("Cân nặng phải là số dương")
    .optional(),
  measurement_notes: z.string().optional(),
  measured_date: z.string().optional(), // ISO date string
});

export type MeasurementCreateInput = z.infer<typeof measurementCreateSchema>;

/**
 * Schema for updating measurement
 */
export const measurementUpdateSchema = measurementCreateSchema.partial();

export type MeasurementUpdateInput = z.infer<typeof measurementUpdateSchema>;

/**
 * Measurement response from API
 */
export interface MeasurementResponse {
  id: string;
  customer_profile_id: string;
  tenant_id: string;
  neck: number | null;
  shoulder_width: number | null;
  bust: number | null;
  waist: number | null;
  hip: number | null;
  top_length: number | null;
  sleeve_length: number | null;
  wrist: number | null;
  height: number | null;
  weight: number | null;
  measurement_notes: string | null;
  is_default: boolean;
  measured_date: string; // ISO date string
  measured_by: string | null;
  created_at: string;
  updated_at: string;
}

// ===== Customer Profile Schemas =====

/**
 * Schema for creating a customer profile
 */
export const customerProfileCreateSchema = z.object({
  full_name: z
    .string()
    .min(2, "Họ tên phải có ít nhất 2 ký tự")
    .max(255, "Họ tên không được quá 255 ký tự"),
  phone: phoneSchema,
  email: z.string().email("Email không đúng định dạng").optional().or(z.literal("")),
  date_of_birth: z.string().optional(), // ISO date string
  gender: z.enum(["Nam", "Nữ", "Khác"]).optional(),
  address: z.string().optional(),
  notes: z.string().optional(),
  initial_measurements: measurementCreateSchema.optional(),
  create_account: z.boolean().optional(), // Checkbox: Tạo tài khoản cho khách hàng
});

export type CustomerProfileCreateInput = z.infer<typeof customerProfileCreateSchema>;

/**
 * Schema for updating customer profile
 */
export const customerProfileUpdateSchema = z.object({
  full_name: z.string().min(2).max(255).optional(),
  phone: phoneSchema.optional(),
  email: z.string().email().optional().or(z.literal("")),
  date_of_birth: z.string().optional(),
  gender: z.enum(["Nam", "Nữ", "Khác"]).optional(),
  address: z.string().optional(),
  notes: z.string().optional(),
});

export type CustomerProfileUpdateInput = z.infer<typeof customerProfileUpdateSchema>;

/**
 * Customer profile response from API
 */
export interface CustomerProfileResponse {
  id: string;
  tenant_id: string;
  user_id: string | null;
  full_name: string;
  phone: string;
  email: string | null;
  date_of_birth: string | null;
  gender: string | null;
  address: string | null;
  notes: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  has_account: boolean; // Computed field
  measurement_count: number; // Computed field
}

/**
 * Customer with measurements response
 */
export interface CustomerWithMeasurementsResponse {
  customer: CustomerProfileResponse;
  measurements: MeasurementResponse[];
  default_measurement: MeasurementResponse | null;
}

/**
 * Paginated customer list response
 */
export interface CustomerListResponse {
  customers: CustomerProfileResponse[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

// ===== API Error Types =====

export interface ApiError {
  detail: string;
}

export interface ValidationError {
  detail: Array<{
    loc: string[];
    msg: string;
    type: string;
  }>;
}

File: backend/src/models/customer.py
"""Pydantic schemas for Customer Profile and Measurements."""

import re
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, EmailStr, Field, field_validator


class CustomerProfileCreateRequest(BaseModel):
    """Schema for creating a new customer profile."""

    full_name: str = Field(..., min_length=2, max_length=255, description="Full name of customer")
    phone: str = Field(..., min_length=10, max_length=20, description="Phone number (VN format)")
    email: EmailStr | None = Field(None, description="Optional email address")
    date_of_birth: date | None = Field(None, description="Date of birth")
    gender: str | None = Field(None, description="Gender: Nam, Nữ, Khác")
    address: str | None = Field(None, description="Full address")
    notes: str | None = Field(None, description="Additional notes")
    create_account: bool = Field(False, description="Create user account for customer if email provided")
    initial_measurements: "MeasurementCreateRequest | None" = Field(
        None, description="Optional initial measurements"
    )

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v: str) -> str:
        """Validate Vietnamese phone format (10-11 digits, starting with 0)."""
        v = v.strip()
        pattern = r"^0[0-9]{9,10}$"
        if not re.match(pattern, v):
            raise ValueError(
                "Số điện thoại không đúng định dạng. Vui lòng nhập số điện thoại Việt Nam (VD: 0901234567)"
            )
        return v

    @field_validator("gender")
    @classmethod
    def validate_gender(cls, v: str | None) -> str | None:
        """Validate gender values."""
        if v is None:
            return v
        allowed = ["Nam", "Nữ", "Khác"]
        if v not in allowed:
            raise ValueError(f"Giới tính phải là một trong: {', '.join(allowed)}")
        return v

    @field_validator("email")
    @classmethod
    def email_lowercase(cls, v: EmailStr | None) -> str | None:
        """Normalize email to lowercase."""
        if v is None:
            return v
        return v.lower()


class CustomerProfileUpdateRequest(BaseModel):
    """Schema for updating customer profile."""

    full_name: str | None = Field(None, min_length=2, max_length=255)
    phone: str | None = Field(None, min_length=10, max_length=20)
    email: EmailStr | None = None
    date_of_birth: date | None = None
    gender: str | None = None
    address: str | None = None
    notes: str | None = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v: str | None) -> str | None:
        """Validate Vietnamese phone format."""
        if v is None:
            return v
        v = v.strip()
        pattern = r"^0[0-9]{9,10}$"
        if not re.match(pattern, v):
            raise ValueError(
                "Số điện thoại không đúng định dạng. Vui lòng nhập số điện thoại Việt Nam (VD: 0901234567)"
            )
        return v

    @field_validator("gender")
    @classmethod
    def validate_gender(cls, v: str | None) -> str | None:
        """Validate gender values."""
        if v is None:
            return v
        allowed = ["Nam", "Nữ", "Khác"]
        if v not in allowed:
            raise ValueError(f"Giới tính phải là một trong: {', '.join(allowed)}")
        return v


class CustomerProfileResponse(BaseModel):
    """Schema for customer profile response."""

    id: uuid.UUID
    tenant_id: uuid.UUID
    user_id: uuid.UUID | None = None
    full_name: str
    phone: str
    email: str | None = None
    date_of_birth: date | None = None
    gender: str | None = None
    address: str | None = None
    notes: str | None = None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime
    has_account: bool = Field(False, description="Whether customer has a user account")
    measurement_count: int = Field(0, description="Number of measurement sets")

    model_config = {"from_attributes": True}


class MeasurementCreateRequest(BaseModel):
    """Schema for creating a new measurement set."""

    neck: Decimal | None = Field(None, ge=20, le=60, description="Vòng cổ (cm)")
    shoulder_width: Decimal | None = Field(None, ge=30, le=60, description="Rộng vai (cm)")
    bust: Decimal | None = Field(None, ge=60, le=180, description="Vòng ngực (cm)")
    waist: Decimal | None = Field(None, ge=40, le=150, description="Vòng eo (cm)")
    hip: Decimal | None = Field(None, ge=60, le=180, description="Vòng mông (cm)")
    top_length: Decimal | None = Field(None, ge=40, le=120, description="Dài áo (cm)")
    sleeve_length: Decimal | None = Field(None, ge=30, le=90, description="Dài tay (cm)")
    wrist: Decimal | None = Field(None, ge=10, le=30, description="Vòng cổ tay (cm)")
    height: Decimal | None = Field(None, ge=100, le=250, description="Chiều cao (cm)")
    weight: Decimal | None = Field(None, ge=30, le=200, description="Cân nặng (kg)")
    measurement_notes: str | None = Field(None, description="Ghi chú số đo đặc biệt")
    measured_date: date | None = Field(None, description="Ngày đo (mặc định: hôm nay)")

    @field_validator(
        "neck",
        "shoulder_width",
        "bust",
        "waist",
        "hip",
        "top_length",
        "sleeve_length",
        "wrist",
        "height",
        "weight",
    )
    @classmethod
    def validate_positive(cls, v: Decimal | None) -> Decimal | None:
        """Validate measurements are positive numbers."""
        if v is not None and v <= 0:
            raise ValueError("Số đo phải là số dương")
        return v


class MeasurementUpdateRequest(BaseModel):
    """Schema for updating measurement set."""

    neck: Decimal | None = Field(None, ge=20, le=60)
    shoulder_width: Decimal | None = Field(None, ge=30, le=60)
    bust: Decimal | None = Field(None, ge=60, le=180)
    waist: Decimal | None = Field(None, ge=40, le=150)
    hip: Decimal | None = Field(None, ge=60, le=180)
    top_length: Decimal | None = Field(None, ge=40, le=120)
    sleeve_length: Decimal | None = Field(None, ge=30, le=90)
    wrist: Decimal | None = Field(None, ge=10, le=30)
    height: Decimal | None = Field(None, ge=100, le=250)
    weight: Decimal | None = Field(None, ge=30, le=200)
    measurement_notes: str | None = None


class MeasurementResponse(BaseModel):
    """Schema for measurement response."""

    id: uuid.UUID
    customer_profile_id: uuid.UUID
    tenant_id: uuid.UUID
    neck: Decimal | None = None
    shoulder_width: Decimal | None = None
    bust: Decimal | None = None
    waist: Decimal | None = None
    hip: Decimal | None = None
    top_length: Decimal | None = None
    sleeve_length: Decimal | None = None
    wrist: Decimal | None = None
    height: Decimal | None = None
    weight: Decimal | None = None
    measurement_notes: str | None = None
    is_default: bool
    measured_date: date
    measured_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerWithMeasurementsResponse(BaseModel):
    """Schema for customer with all measurements."""

    customer: CustomerProfileResponse
    measurements: list[MeasurementResponse] = []
    default_measurement: MeasurementResponse | None = None
